### ISSUE TITLE
Version 10.0.1 introduces bug deserializing double -52.3698169

### ISSUE THREAD

**Issue by RobDaPraia** [JamesNK/Newtonsoft.Json#1255]
After upgrading to version 10.0.1 the double -52.3698169 is not deserialized correctly. This issue did not happen in version 9.0.1. Other values like -75.3062394 are desrialized correctly.

You can run the following unit test to check this issue

```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json;

namespace NewtonJsonTest
{

    // After NuGet update to NewtonJson 10.0.1 we suddenly saw an error when deserializing/serializing a GeoJson string
    // Therefore these tests to find the reason and check that Newton Json is working correctly
    // Seems that there is a bug in NewtonJson, see test DoubleNotOkTest
    //
    // Note: no errors with NewtonJson version 9.0.1


    [TestClass]
    public class UnitTests
    {

        [TestMethod]
        public void DoubleOkTest()
        {
            var expectedJson = "-75.3062394";
            double longitude = -75.3062394;
            var json = JsonConvert.SerializeObject(longitude);
            Assert.AreEqual(expectedJson, json, "Assert 1");

            var doubleFromJson = JsonConvert.DeserializeObject<double>(json);

            Assert.AreEqual(-75.3062394, doubleFromJson, "Assert 2");
            var serializedDouble = JsonConvert.SerializeObject(doubleFromJson);
            Assert.AreEqual(expectedJson, serializedDouble, "Assert 3");
        }


        [TestMethod]
        public void DoubleNotOkTest()
        {
            var expectedJson = "-52.3698169";
            double longitude = -52.3698169;
            var json = JsonConvert.SerializeObject(longitude);
            Assert.AreEqual(expectedJson, json, "Assert 1");

            var doubleFromJson = JsonConvert.DeserializeObject<double>(json);

            // Wrong deserializing,  doubleFromJson	-52.369816900000004	double
            Assert.AreEqual(-52.3698169, doubleFromJson, "Assert 2");
            var serializedDouble = JsonConvert.SerializeObject(doubleFromJson);
            Assert.AreEqual(expectedJson, serializedDouble, "Assert 3");
        }
    }


}
```

**Comment by TylerBrinkley** [comment]
@pieceofsummer any ideas?

**Comment by FurkanKambay** [comment]
I found that it happens because of this if statement here:
https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Utilities/ConvertUtils.cs#L1109

```C#
// val = 15094572024083270656 at this point.
// for -75.3062394, val = 10852772854416427657
if ((val & (1 << 10)) != 0)
{
    // IEEE round to even
    ulong tmp = val + ((1UL << 10) - 1 + ((val >> 11) & 1));
    if (tmp < val)
    {
        // overflow
        tmp = (tmp >> 1) | 0x8000000000000000;
        exp++;
    }
    val = tmp;
}
```

**Comment by TylerBrinkley** [comment]
Hmm, it looks the same as https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/number.cpp#L534

```cpp
// round & scale down
if ((UINT32)val & (1 << 10))
{
    // IEEE round to even
    UINT64 tmp = val + ((1 << 10) - 1) + (((UINT32)val >> 11) & 1);
    if (tmp < val) {
        // overflow
        tmp = (tmp >> 1) | I64(0x8000000000000000);
        exp += 1;
    }
    val = tmp;
}
```

**Comment by JamesNK** [comment]
This needs to get fixed. I don't have the knowledge to make changes here.

**Comment by JamesNK** [comment]
I wrote a test app for comparing the result from double.Parse vs Json.NET double parsing using random text and there are A LOT of incorrect values.

https://gist.github.com/JamesNK/4e6bee576196be016fdd31a3dbe0880e

Small sample:
-1984665458.1609920270
-1046736634.198882997
1455973205.597988487
1745567008.1420804274
-2127704185.866700058
-1612079937.189740309
1639795656.1972154380
1644154510.338570954
2084446913.392945177
-1861273728.928892732
146874326.2121974088
-267181755.1674678778
2086908639.1570776785
-1347750616.336657409
-2067263735.1014753590
1656147426.694348697
2132752265.583266712
-1065983758.1643603483
1530373023.246784568
1607769239.845104577
1287975122.888343695
-2124254827.2130717051
-129246428.1179858226
-1476792472.1003645691
2147348374.494998816
-104282459.1315922369
-337262284.240940541
-598760065.1588140138
-542180617.1638948321
489187934.1210820979
-1990110285.112074497
454106690.1930053534
1076927570.511446835
1914119494.1606444142
2023741479.1223870582

**Comment by pieceofsummer** [comment]
Oh, it appears that different rounding rules are used between frameworks!

Take a simple program:
```
using System;
using System.Globalization;

class Program
{
	static void Main(string[] args)
	{
		var str = "-52.3698169";
		var value = double.Parse(str, CultureInfo.InvariantCulture);
		Console.WriteLine("0x{0:X16}", BitConverter.DoubleToInt64Bits(value));
	}
}
```

On classic .NET it produces `0xC04A2F5629018109`, while on Mono / .NET Core – `0xC04A2F562901810A`.

**Comment by TylerBrinkley** [comment]
Great work @pieceofsummer!

The only thing that I found was that the C++ version only considers the first 18 digits whereas the C# version considers the first 19 digits.

**Comment by pieceofsummer** [comment]
@TylerBrinkley not yet. It is really a pain to get to`internalcall` method implementation from a disassembler. Might need to dive into WinDbg :(

**Comment by TylerBrinkley** [comment]
@pieceofsummer bummer. It might be easier to just guess how it rounds. We should be able to easily determine which rounding scheme the framework uses by using `double.Parse` on a problem value.

**Comment by pieceofsummer** [comment]
Oh, it even depends on architecture, because 64-bit classic .NET runtime also produces `0xC04A2F562901810A` for the code above (TestConsole was running in 32-bit mode by default).

**Comment by JamesNK** [comment]
I did some Googling:

http://stackoverflow.com/questions/4018895/why-does-math-exp-give-different-results-between-32-bit-and-64-bit-with-same-in
https://github.com/dotnet/roslyn/issues/4221

https://github.com/dotnet/roslyn/blob/614299ff83da9959fa07131c6d0ffbc58873b6ae/src/Compilers/Core/Portable/RealParser.cs

**Comment by JamesNK** [comment]
This looks like it is pretty complicated and doesn't have a simple answer. I want to get a new release out soon so I might revert back to double.TryParse for now.

**Comment by FurkanKambay** [comment]
@RobDaPraia I just tried `-52.3698169` in 9.0.1 and it threw the same error. Are you sure?

@JamesNK I have been googling too. I have one or two ideas and trying to get my head around them. But I think there's not really much that can be done?
Also, the file you sent says:
>... This code does not support a leading sign character as that is not part of the C# or VB floating-point literal lexical syntax.
So this won't be our solution either.

**Comment by RobDaPraia** [comment]
@FurkanKambay

Yes, all my unit tests worked with 9.0.1 and started to throw errors only after upgrade to version 10.0.1 I had also just upgrades to VS2017, not sure if that is relevant

I have attached my test solution with the unit test, check if that works for you with version 9.0.1

![image](https://cloud.githubusercontent.com/assets/3207395/24320344/c4e1392a-1133-11e7-9038-be70df2f0c63.png)

[NewtonJsonTest.zip](https://github.com/JamesNK/Newtonsoft.Json/files/869820/NewtonJsonTest.zip)

**Comment by FurkanKambay** [comment]
@RobDaPraia You're right, it does work.

**Comment by JamesNK** [comment]
10.0.2 uses `double.TryParse` again. It will parse doubles correctly again.

**Comment by RobDaPraia** [comment]
Thanks for the solution :)
